### PR TITLE
[WIP] Support cross-compilation to Scala Native, fix #1112

### LINKED ISSUE #1112: Add support for Scala Native
It is early days for Scala Native (https://github.com/scala-native/scala-native), but support would be be very helpful.

### PR DESCRIPTION
**Pull request by shadaj**
This PR needs a lot more work, but wanted to get it out there for an initial review and feedback. I have tested the Scala Native build of ScalaTest with another project (https://github.com/team846/potassium), and the basic functionality is working (with one Scala Native optimization phase disabled that currently crashes).

Right now, the PR depends on unmerged PRs from ScalaTest's dependencies, such as ScalaCheck. Also right now the unit tests have not been ported to Scala Native, but I will be working on making them work next.

cc @densh @Duhemm

### CODE REVIEW

File: project/scalatest.scala [hunk 1]
```diff
@@ -199,6 +201,11 @@ object ScalatestBuild extends Build {
       "org.scala-js" %% "scalajs-test-interface" % "0.6.17"
     )
 
+  def scalatestNativeLibraryDependencies =
+    Seq(
+      "org.scala-native" % "test-interface_native0.3_2.11" % "0.3.1"
```
Comment (MasseGuillaume) [inline project/scalatest.scala]: can you use %%% here ?

Comment (MasseGuillaume) [inline project/scalatest.scala]: see https://github.com/scala-native/sbt-crossproject#cross-compiling-scalajs-jvm-and-native

Comment (shadaj) [inline project/scalatest.scala]: Fixed!

**Comment by shadaj** [comment]
@bvenners I've finished up some more testing with my own projects and I think this PR is ready for a first review!

**Comment by bvenners** [comment]
Cool. Thanks @shadaj for submitting this. Sorry we haven't had many cycles to review it yet. It is a rather large PR. I wonder if we could set up a time to walk us through it over a hangout?

**Comment by shadaj** [comment]
Yeah, going through it over a hangout sounds like a great idea!

**Comment by shadaj** [comment]
@bvenners I just fixed some remaining compilation issues for generated sources, so now everything should build from a fresh clone!

**Comment by fsat** [comment]
@shadaj - I'd like to continue our discussion from https://github.com/fsat/k8s-cli/issues/1#issuecomment-328362219 in this PR if that's ok.

I've tried with the latest commit, the compile error is no longer present, however I have the following error when running `sbt publishLocal`:

```
[info] 	published ivy to /Users/felixsatyaputra/.ivy2/local/org.scalactic/scalactic_2.11/3.0.1/ivys/ivy.xml
[info] Wrote /Users/felixsatyaputra/workspace/typesafe-fsat/scalatest/target/scala-2.11/scalatest-app_2.11-3.0.1.pom
sbt.ResolveException: unresolved dependency: org.scalacheck#scalacheck_2.11;1.14.0-native-SNAPSHOT: not found
```

I've used `sbt publishLocal` on my local clone of `scalacheck`, and I've noticed it's published as `1.14.0-SNAPSHOT`.

This is different in the version required by this PR. I noticed that `project/scalatest.scala` has `scalacheckVersion` specified as `1.14.0-native-SNAPSHOT`.

How do you publish the local `scalacheck` snapshot required for this PR?

**Comment by cheeseng** [comment]
@shadaj I am trying this on Ubuntu, but got the following error when trying to run sbt scalatestNative/test: 

```
scala.collection.parallel.CompositeThrowable: Multiple exceptions thrown during a parallel computation: java.util.NoSuchElementException: key not found: Block(Local(107),Vector(),UnrolledBuffer(Let(Local(538),Classalloc(Top(scala.MatchError))), Let(Local(539),Box(Class(Top(java.lang.Character)),Local(Local(79),Char))), Let(Local(540),Call(Function(Vector(Class(Top(scala.MatchError)), Class(Top(java.lang.Object))),Unit),Global(Member(Top(scala.MatchError),init_java.lang.Object),Ptr),Vector(Local(Local(538),Class(Top(scala.MatchError))), Local(Local(539),Class(Top(java.lang.Character)))),None)), Throw(Local(Local(538),Class(Top(scala.MatchError))),None)),false)
scala.collection.MapLike$class.default(MapLike.scala:228)
scala.collection.AbstractMap.default(Map.scala:58)
scala.collection.MapLike$class.apply(MapLike.scala:141)
scala.collection.AbstractMap.apply(Map.scala:58)
scala.scalanative.optimizer.pass.GlobalBoxingElimination.isDominatedBy$1(GlobalBoxingElimination.scala:40)
scala.scalanative.optimizer.pass.GlobalBoxingElimination.scala$scalanative$optimizer$pass$GlobalBoxingElimination$$canReuse$1(GlobalBoxingElimination.scala:45)
scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:62)
scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:54)
scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:132)
scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:131)
.
.
.

java.lang.NullPointerException
scala.collection.parallel.AugmentedIterableIterator$class.map2combiner(RemainsIterator.scala:119)
scala.collection.parallel.immutable.ParHashMap$ParHashMapIterator.map2combiner(ParHashMap.scala:79)
scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1057)
scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1054)
scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
.
.
.
	at scala.collection.parallel.package$$anon$1.alongWith(package.scala:86)
	at scala.collection.parallel.Task$class.mergeThrowables(Tasks.scala:86)
	at scala.collection.parallel.ParIterableLike$Map.mergeThrowables(ParIterableLike.scala:1054)
	at scala.collection.parallel.Task$class.tryMerge(Tasks.scala:72)
	at scala.collection.parallel.ParIterableLike$Map.tryMerge(ParIterableLike.scala:1054)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:190)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
	at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[error] (scalatestNative/nativetest:nativeOptimizeNIR) scala.collection.parallel.CompositeThrowable: Multiple exceptions thrown during a parallel computation: java.util.NoSuchElementException: key not found: Block(Local(107),Vector(),UnrolledBuffer(Let(Local(538),Classalloc(Top(scala.MatchError))), Let(Local(539),Box(Class(Top(java.lang.Character)),Local(Local(79),Char))), Let(Local(540),Call(Function(Vector(Class(Top(scala.MatchError)), Class(Top(java.lang.Object))),Unit),Global(Member(Top(scala.MatchError),init_java.lang.Object),Ptr),Vector(Local(Local(538),Class(Top(scala.MatchError))), Local(Local(539),Class(Top(java.lang.Character)))),None)), Throw(Local(Local(538),Class(Top(scala.MatchError))),None)),false)
[error] scala.collection.MapLike$class.default(MapLike.scala:228)
[error] scala.collection.AbstractMap.default(Map.scala:58)
[error] scala.collection.MapLike$class.apply(MapLike.scala:141)
[error] scala.collection.AbstractMap.apply(Map.scala:58)
[error] scala.scalanative.optimizer.pass.GlobalBoxingElimination.isDominatedBy$1(GlobalBoxingElimination.scala:40)
[error] scala.scalanative.optimizer.pass.GlobalBoxingElimination.scala$scalanative$optimizer$pass$GlobalBoxingElimination$$canReuse$1(GlobalBoxingElimination.scala:45)
[error] scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:62)
[error] scala.scalanative.optimizer.pass.GlobalBoxingElimination$$anonfun$onInsts$1$$anonfun$apply$1.isDefinedAt(GlobalBoxingElimination.scala:54)
[error] scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:132)
[error] scala.collection.TraversableOnce$$anonfun$collectFirst$1.apply(TraversableOnce.scala:131)
[error] .
[error] .
[error] .
[error] 
[error] java.lang.NullPointerException
[error] scala.collection.parallel.AugmentedIterableIterator$class.map2combiner(RemainsIterator.scala:119)
[error] scala.collection.parallel.immutable.ParHashMap$ParHashMapIterator.map2combiner(ParHashMap.scala:79)
[error] scala.collection.parallel.ParIterableLike$Map.leaf(ParIterableLike.scala:1057)
[error] scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
[error] scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
[error] scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
[error] scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
[error] scala.collection.parallel.ParIterableLike$Map.tryLeaf(ParIterableLike.scala:1054)
[error] scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
[error] scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
[error] .
[error] .
[error] .
[error] Total time: 133 s, completed Oct 6, 2017 5:52:18 PM

```
Any idea what could be wrong?  I have setup my environment following the setup guide here: http://www.scala-native.org/en/latest/user/setup.html

```
$ sudo apt install clang libunwind-dev
$ sudo apt install libgc-dev libre2-dev # optional
```
Thanks.

**Comment by MasseGuillaume** [comment]
@densh any idea what is wrong here ^ ?

**Comment by shadaj** [comment]
This is caused by https://github.com/scala-native/scala-native/issues/1009. If you run the tests through the command `scalatestTestNative/test`, things should work since for that command I disabled the phase of optimization that crashes.

**Comment by cheeseng** [comment]
@shadaj scalatestTestNative/test seems to fix the problem!  now I got: 

```
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
```
in the linking stage, which I remember in the hangout you mentioned it needs more size.  I'll give it another shot tomorrow with more memory.

Thanks!

**Comment by cheeseng** [comment]
@shadaj It linked and runs now with 8G of heap size, looks great: 

```
[error] Failed: Total 5566, Failed 51, Errors 0, Passed 5515, Ignored 14, Canceled 6, Pending 4
[error] Failed tests:
[error] 	org.scalatest.tools.ArgsParserSpec
[error] 	org.scalatest.words.BeWordSpec
[error] 	org.scalatest.time.SpanSugarSpec
[error] 	org.scalatest.time.SpanSpec
[error] 	org.scalatest.events.deprecated.DeprecatedLocationFunctionSuiteProp
[error] 	org.scalatest.words.NotWordSpec
[error] (scalatestTestNative/nativetest:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 499 s, completed Oct 7, 2017 12:06:44 PM
```

I'll dig into the failing test to understand why, but this looks really neat!

**Comment by cheeseng** [comment]
By the way, when I tried to run scalacticTestNative I got: 

```
[error] cannot link: @java.lang.Thread::getStackTrace_scala.scalanative.runtime.ObjectArray
[error] unable to link
[error] (scalacticTestNative/nativetest:nativeLinkNIR) unable to link
[error] Total time: 92 s, completed Oct 7, 2017 12:10:53 PM
```

Not sure the reason, I'll focus on the failing tests in scalatest tests first.

**Comment by shadaj** [comment]
@cheeseng I just added a commit that fixes Scalactic tests not running, by having Scala Native not fail the compile if we are using stubbed Java library methods (in this case `getStackTrace`).

**Comment by cheeseng** [comment]
@shadaj @densh @MasseGuillaume  

Some finding of the failing tests: 

org.scalatest.tools.ArgsParserSpec
This seems to be due to java.lang.Double.parseDouble returns 0 in case of invalid data, and seems like the following new commit in scala-native should fix it: 

https://github.com/scala-native/scala-native/commit/8ed99b3a74d8912e62988cfb9ef90d9730bc860d

The more interesting one is in org.scalatest.words.BeWordSpec, which I think something is wrong with String's replaceAllLiterally method implementation.

The following code: 

```
class MyFile {
  def isWritable: Boolean = true
}

val myFile = new MyFile

println("###toString: " + myFile.toString)
println("###toString + \"\": " + myFile.toString + "")
println("###myFile + \"\": " + "{0} was not writable".replaceAllLiterally("{0}", myFile.toString + ""))
```

is giving the following output: 

```
###toString: org.scalatest.words.BeWordSpec$$anonfun$1$$anonfun$apply$mcV$sp$24$MyFile$38@9de1360
###toString + "": org.scalatest.words.BeWordSpec$$anonfun$1$$anonfun$apply$mcV$sp$24$MyFile$38@9de1360
###myFile + "": org.scalatest.words.BeWordSpec was not writable
```

Not sure why the $$ part of the toString is removed by the replaceAllLiterally.

@bvenners Do you have any idea how the following numbers were worked out?

https://github.com/scalatest/scalatest/blame/6e03d4d77462d0f1ab2fc211c22c03c4150b9a30/scalatest-test/src/test/scala/org/scalatest/time/SpanSugarSpec.scala#L171

**Comment by cheeseng** [comment]
Hi @shadaj , 

Just to inform that we're currently working on few build related PR to clean things up, we shall come back to this after it is completed.

Looking forward to bring scala-native build in!

Thanks!

**Comment by cheeseng** [comment]
Another possible bug in scala-native Double's round method implementation: 

```
val a: Double = 4228438087383875356545203991520.000000
a.round
```

In JVM, the code above returns: 

```
9223372036854775807
```

but in scala-native, it returns: 

```
-9223372036854775808.000000
```

Perhaps I shall report the detected bugs in scala-native github issues?

**Comment by shadaj** [comment]
@cheeseng Interesting! Yeah, that should probably be filed as an issue for scala-native, since it's resulting in different behavior compared to the JVM.

**Comment by cheeseng** [comment]
@shadaj @densh @MasseGuillaume 

Not sure when this error gets in, I am trying to pull latest changes in upstream 3.1.x into @shadaj branch, but before doing so I am trying to run scalatestTestNative/test but now got this error when running: 

```
java.net.SocketException: Socket is closed
	at java.net.Socket.setSoTimeout(Socket.java:1137)
	at scala.scalanative.sbtplugin.testinterface.ComRunner.receive(ComRunner.scala:102)
	at scala.scalanative.sbtplugin.testinterface.ScalaNativeRunner.receive(ScalaNativeRunner.scala:41)
	at scala.scalanative.sbtplugin.testinterface.ScalaNativeTask.receive$1(ScalaNativeTask.scala:32)
	at scala.scalanative.sbtplugin.testinterface.ScalaNativeTask.execute(ScalaNativeTask.scala:43)
	at sbt.TestRunner.runTest$1(TestFramework.scala:76)
	at sbt.TestRunner.run(TestFramework.scala:85)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] Could not run test org.scalatest.EveryShouldContainInOrderElementsOfLogicalAndSpec: java.net.SocketException: Socket is closed
java.lang.NullPointerException
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:78)
	at sbt.JUnitXmlTestsListener$TestSuite$$anonfun$2.apply(JUnitXmlTestsListener.scala:71)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:32)
	at scala.collection.mutable.ListBuffer.foreach(ListBuffer.scala:45)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at sbt.JUnitXmlTestsListener$TestSuite.stop(JUnitXmlTestsListener.scala:71)
	at sbt.JUnitXmlTestsListener.writeSuite(JUnitXmlTestsListener.scala:180)
	at sbt.JUnitXmlTestsListener.endGroup(JUnitXmlTestsListener.scala:161)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestRunner$$anonfun$run$4.apply(TestFramework.scala:90)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at sbt.TestFramework$$anonfun$safeForeach$1.apply(TestFramework.scala:107)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at sbt.TestFramework$.safeForeach(TestFramework.scala:107)
	at sbt.TestRunner.safeListenersCall(TestFramework.scala:96)
	at sbt.TestRunner.run(TestFramework.scala:90)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
	at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
	at sbt.TestFunction.apply(TestFramework.scala:207)
	at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] java.lang.NullPointerException
```

Do you have any clue what could be causing that?  As far as I can remember the last time I ran it didn't have this issue.  I am running on Ubuntu Linux.

Thanks.

**Comment by Duhemm** [comment]
@cheeseng The `SocketClosedException` usually means that the remote test runner crashed (with a segfault or a division by zero for instance). 

Maybe that there are new tests that do division by zero or null dereference?

**Comment by cheeseng** [comment]
Hi @Duhemm , following your hint I tried to narrow the failure down, but encountered a really weird behavior where a single debugging println statement is avoiding the crash, you can try to checkout my following branch: 

https://github.com/cheeseng/scalatest/tree/scala-native-debug

After checking out the branch you can try: 

sbt ++2.11.11 "scalatestTestNative/test-only org.scalatest.EveryShouldContainAllElementsOfSpec"

It should compile, link and when it runs the test it'll crashed with Socket is closed error.

If you then uncomment this println: 

https://github.com/cheeseng/scalatest/blob/scala-native-debug/scalatest/src/main/scala/org/scalatest/enablers/InspectorAsserting.scala#L677

and try to rerun, it runs without the crash and you'll get a passing test!

To be honest, I have no idea what's going on, a println can avoid segfault?

**Comment by Duhemm** [comment]
@cheeseng Sorry you encountered such a problem. That's probably a bug in Scala Native... Does it still happen with the current master? All the instruction to build Scala Native on your machine are available at http://www.scala-native.org, in case you want to try. I'll try it when I get some time...

**Comment by shadaj** [comment]
@cheeseng I went ahead and merged in 3.1.x so that we have a common base to debug. I'm also getting the same error and am working on debugging.

**Comment by shadaj** [comment]
@cheeseng so after disabling the crashing test, it looks like something fundamentally broke with `contain`, as there are now 900+ failing tests related to the `should contain ...` pattern. Do you know of any changes that might have caused this?

**Comment by cheeseng** [comment]
Sorry @Duhemm and @shadaj for replying late, I just came back from leaves.

@Duhemm I tried current master (built it locally and use 0.4.0-SNAPSHOT), it seems to have the same problem.  In my humble opinion, it would be interesting and useful to find the real reason behind the crash, since we have a case that can reproduce it consistently.

@shadaj Bill did pulled in quite a number of PRs into 3.1.x recently, I'll take a look at the failing tests in your current branch and see if I can figure out the reason they are failing.

**Comment by cheeseng** [comment]
@shadaj when I tried to run your current branch (with one suite disabled), I still get the same error, probably some other tests are hitting the same problem I suppose?

File: project/plugins.sbt [hunk 2]
```diff
@@ -4,4 +4,10 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-osgi" % "0.7.0")
 
 addSbtPlugin("org.scala-js" % "sbt-scalajs" % "0.6.20")
 
+addSbtPlugin("org.scala-native" % "sbt-crossproject"         % "0.2.0")
```
Comment (MasseGuillaume) [inline project/plugins.sbt]:
```
addSbtPlugin("org.scala-js"       % "sbt-scalajs"              % "0.6.21")
addSbtPlugin("org.portable-scala" % "sbt-crossproject"         % "0.3.0")
addSbtPlugin("org.portable-scala" % "sbt-scalajs-crossproject" % "0.3.0")
```

**Pull request merged**